### PR TITLE
feat(tui): show inline colored diffs for patch_apply tool

### PR DESCRIPTION
**Pull request by griffinwork40**
## Summary

`patch_apply` now shows inline colored diffs in the TUI, matching the rendering that `edit_file` already has. Previously, `patch_apply` embedded the diff as a string inside JSON `content` -- the user saw raw JSON instead of a formatted diff block.

## Problem

The `patch_apply` tool returned its diff as a plain string buried inside `JSON.stringify`'d content with no `render.diff` sidecar. The entire TUI rendering pipeline (`tool.diff` provider event -> `tool_diff` stream chunk -> `formatDiffBlock`) already existed for `edit_file` but `patch_apply` never plugged into it.

## Changes

**`src/utils/diff.ts`** -- Added optional `filePath` field to `DiffHunk` interface. Multi-file tools stamp each hunk with its source file path so the TUI can render file-boundary headers. Backward-compatible: single-file diffs don't set it.

**`src/agent/tools/handlers/patch-apply-engine.ts`** -- Refactored `formatUnifiedDiff` into `computeFileDiff` which returns both the string representation AND the structured `DiffPayload` (previously discarded). Added `structuredDiff` field to `PatchApplyResult` carrying a merged `DiffPayload` across all files.

**`src/agent/tools/handlers/patch-apply.ts`** -- Attached `render: { diff }` to the `ToolResult` when a structured diff is available, enabling the existing provider sidecar pipeline.

**`src/cli/commands/interactive/tool-lane-format-diff.ts`** -- Added file-path header rendering at file boundaries in `formatDiffBlock`. When consecutive hunks have different `filePath` values, a dim file-path line is emitted before the `@@` header.

## What it looks like

Multi-file patch rendering:
```
  +5 -3 across 4 hunks
  src/a.ts
  @@ -10,3 +10,5 @@
    context
  - old line
  + new line
  src/b.ts
  @@ -1,2 +1,3 @@
  + added line
```

Single-file patches render identically to `edit_file`.

## Testing

- 7 new tests across `patch-apply.test.ts` (4 tests: render.diff on apply, multi-file filePath stamps, dry_run, empty changes) and `tool-lane-format.test.ts` (3 tests: file-path headers between files, same-file dedup, absence for single-file diffs)
- All 199 tests pass across the 4 affected test suites
- `pnpm lint` passes
- No provider or stream-consumer changes needed -- both Anthropic and OpenAI providers already emit `tool.diff` when `render.diff` is present

### CODE REVIEW

**Comment by griffinwork40** [comment]
**Wave 1.5 verification manifest:**

| # | Type | Claim | Status | Evidence |
|---|---|---|---|---|
| 1 (F1) | citation | `partial_failure` return at lines ~256-261 omits `structuredDiff` | **verified** | `git show 6bd52ebe:...engine.ts` lines 256-261: `{ status: 'partial_failure', diff: combinedDiff, files_changed: filesChanged, errors: writeErrors }` -- no `structuredDiff` |
| 1 (F1) | citation | second `partial_failure` at lines ~315-320 also omits | **verified** | lines 315-320: `{ status: 'partial_failure', diff: combinedDiff, files_changed: filesChanged, errors: renameErrors }` -- no `structuredDiff` |
| 2 (F2) | citation | `computeFileDiff(change.path, ...)` uses raw input path | **verified** | line 165: `computeFileDiff(change.path, originalContent, newContent)`, line 167: `filePath: change.path` |
| 3 (F3) | citation | `palette.dim(hunk.filePath)` with no sanitizer | **verified** | `tool-lane-format-diff.ts` line 226: `items.push({ kind: 'header', text: palette.dim(hunk.filePath) })` |
| 4 (F5) | absence | `partial_failure` absent from `patch-apply.test.ts` render block | **confirmed-absent** | `git grep -n 'partial_failure' 6bd52ebe -- src/agent/tools/handlers/patch-apply.test.ts` = exit 1 (zero matches in entire file) |
| 5 (api-compat) | absence | `formatUnifiedDiff` has no external importers | **confirmed-absent** | `git grep` returns zero matches in production src |
| 6 (F3) | absence | `sanitizeLabel` not imported in `tool-lane-format-diff.ts` | **confirmed-absent** | `git grep -n 'sanitizeLabel' 6bd52ebe -- src/cli/commands/interactive/tool-lane-format-diff.ts` = zero matches |

All citations verified. No fabricated or false-absent claims. Zero findings dropped or downgraded.

Now Wave 2 -- synthesis. No findings at `critical` or `high`, no `blocking` overrides, so no shadow-verify is required.

---

## Review: PR #1497 — feat(tui): show inline colored diffs for patch_apply tool

**Branch:** `afk/show-patch-apply-diff` → `main`
**Ref:** `6bd52ebe`
**Regime:** light (+195/-9, 6 files, feature)
**Stated intent:** Present — PR title + body.

---

### Findings

**1.** `low` · `blocking:false` · confidence:high · **correctness** · `src/agent/tools/handlers/patch-apply-engine.ts:256-261, 315-320` · ref:`6bd52ebe` · file-state

Both `partial_failure` return paths omit `structuredDiff`, silently dropping the already-computed structured diff for files that succeeded Phase 1.

```typescript
return {
  status: 'partial_failure',
  diff: combinedDiff,
  files_changed: filesChanged,
  errors: writeErrors,   // or renameErrors
};
```

**Suggestion:** Add `structuredDiff,` to both `partial_failure` return objects. The handler spreads it conditionally on truthiness so no call-site change is needed.

---

**2.** `low` · `blocking:false` · confidence:high · **correctness** · `src/agent/tools/handlers/patch-apply-engine.ts:165-167` · ref:`6bd52ebe` · file-state

`computeFileDiff` and the hunk `filePath` stamp both receive `change.path` (the raw model-provided string, potentially relative) rather than `resolvedPath` (the canonicalized absolute path). The TUI file-boundary header will display whatever string the model sent, not the canonical filesystem path.

```typescript
const fileDiff = computeFileDiff(change.path, originalContent, newContent);
planned.push({ ..., filePath: change.path, ... });
```

**Suggestion:** This is likely intentional -- the diff headers (`--- a/src/foo.ts`) use the user-facing short path, and `filePath` serves the same rendering purpose. If so, no change needed. If canonical paths are preferred, replace `change.path` with `resolvedPath`.

---

**3.** `nit` · `blocking:false` · confidence:medium · **correctness** · `src/cli/commands/interactive/tool-lane-format-diff.ts:226` · ref:`6bd52ebe` · file-state

The file-path header passes `hunk.filePath` through `palette.dim()` without the `stripAnsi` + C0-scrub pipeline that diff body lines receive via `colorDiffLine`. Not a practical risk (file paths are model-supplied strings unlikely to contain ANSI sequences), but inconsistent with the existing sanitization posture for the diff render path.

```typescript
items.push({ kind: 'header', text: palette.dim(hunk.filePath) });
```

**Suggestion:** Wrap with `sanitizeLabel(hunk.filePath)` before `palette.dim(...)` for consistency.

---

**4.** `low` · `blocking:false` · confidence:high · **test-coverage** · `src/agent/tools/handlers/patch-apply.test.ts:278-289` · ref:`6bd52ebe` · file-state

The "omits render.diff when no changes" test only covers the `changes: []` early-return path, not the path where files are supplied but produce identical before/after content (`computeLineDiff` returns `null`, `mergedHunks.length === 0`).

**Suggestion:** Add a test with a non-empty `changes` array where content is unchanged, verifying `render` is undefined.

---

**5.** `low` · `blocking:false` · confidence:high · **test-coverage** · `src/agent/tools/handlers/patch-apply.test.ts` · ref:`6bd52ebe` · file-state

No test in the `render.diff sidecar` block covers the `partial_failure` paths. Grep confirms zero matches for `partial_failure` in the entire test file.

**Suggestion:** Add a test triggering `partial_failure` and asserting on the `render` field.

---

**6.** `low` · `blocking:false` · confidence:high · **test-coverage** · `src/cli/commands/interactive/tool-lane-format.test.ts:1310-1312` · ref:`6bd52ebe` · file-state

The multi-file header test verifies `src/a.ts` header comes before its `@@` line but does not verify `src/b.ts` header comes after `src/a.ts`'s hunk content.

**Suggestion:** Assert `lines.indexOf('  src/b.ts') > aHunkIdx` for full ordering verification.

---

### Dimensions with no issues

- **security** — no issues found. `structuredDiff` is structurally excluded from model-facing `tool_result` content. `filePath` values pass through the existing `palette.dim()` render path. No injection surface introduced.
- **api-compat** — no issues found. `DiffHunk.filePath` is an optional field addition (backward-compatible). `PatchApplyResult.structuredDiff` is optional. The renamed `computeFileDiff` was module-private (`formatUnifiedDiff` had no export, zero external importers confirmed by grep).
- **perf-observability** — no issues found. The merge loop is O(total hunks), matching existing `combinedDiff` construction. `WeakMap` memo cache is correctly keyed per fresh `DiffPayload` allocation.
- **spec-compliance** — stated intent fully met. `patch_apply` now emits `render: { diff }` for TUI rendering on both `applied` and `dry_run` paths. Multi-file file-boundary headers work. No scope creep.

---

### What was not checked

- Citations verified inline against branch HEAD `6bd52ebe`.
- Stated intent: PR #1497 title+body -- spec-compliance assessed.
- Did not run tests or build.
- Did not review Telegram or non-REPL surfaces for `render.diff` handling.
- Shadow-verify not triggered (no critical/high/overridden findings).

---

**Decision: MERGE — 0 blocking (4 low, 1 nit).**

Clean feature change. The existing `edit_file` diff rendering pipeline is correctly reused. All 6 findings are low-severity suggestions -- the `partial_failure` path omission (Finding 1) is the most worthwhile follow-up.

---
_🤖 Posted by `agent-afk /review --post github`_

**Review by griffinwork40 — COMMENTED** [review]
🤖 Automated review (hourly sweep), generated by the /review tool — a maintainer will follow up.

## Decision: MERGE — 0 blocking (2 low, 3 nit)

**PR #1497** — feat(tui): show inline colored diffs for patch\_apply tool
**Ref:** \`6bd52ebe48\` · 6 files, ~216 lines · feature · light regime

---

### Findings

**low · blocking:false · high-confidence · correctness · \`patch-apply-engine.ts:~257,~315\`**
\`structuredDiff\` is not propagated on the two \`partial_failure\` return paths in \`applyPatch\`. The variable is in scope (computed at line ~186) but omitted from both partial-failure returns. This means the TUI diff rendering silently disappears when a multi-file patch partially succeeds — even though the structured diff was already computed. Not a regression (old behavior was also raw JSON on this path), but a consistency gap given the PR's intent.
**Suggestion:** Include \`structuredDiff\` in both \`partial_failure\` return objects.

**low · blocking:false · high-confidence · test-coverage · \`patch-apply.test.ts\`**
No test covers \`render.diff\` behavior on the \`partial_failure\` path (zero occurrences of "partial" in the new test suite). Given the gap in Finding 1, this masks the inconsistency.
**Suggestion:** Add a test that induces a partial failure and asserts on \`render.diff\` presence/absence.

**nit · blocking:false · high-confidence · api-compat · \`src/utils/diff.ts:36–42\`**
\`DiffHunk.filePath\` is optional and additive — no breaking change for existing consumers. Clean.

**nit · blocking:false · medium-confidence · correctness · \`tool-lane-format-diff.ts:219+\`**
File-path headers are rendered as \`kind: 'header'\` items exempt from the body-line cap. A 100-file patch would emit 100 uncapped header lines. Benign at current scale; worth noting for future large-patch scenarios.

**nit · blocking:false · high-confidence · test-coverage**
All three new TUI tests accurately capture the intended file-path header contract: multi-file boundaries, same-file dedup, and single-file-no-filePath absence. No issues.

---

### Spec-Compliance

The PR satisfies its stated intent:
- ✅ \`patch_apply\` shows inline colored diffs via \`render.diff\` sidecar
- ✅ Matches \`edit_file\` rendering (same \`formatDiffBlock\` code path)
- ✅ Optional \`filePath\` field on \`DiffHunk\` — backward-compatible
- ✅ File-boundary headers rendered in TUI for multi-file diffs
- ✅ 7 new tests across engine and TUI formatting

One consistency gap: \`partial_failure\` paths do not propagate \`structuredDiff\`, so TUI diffs are absent on partial success. Not a regression — the old code had no TUI diff at all — but a reachable gap under the PR's scope.

### What was not checked
- Citations verified inline against branch HEAD \`6bd52ebe48\`.
- Stated intent: PR title + body — spec-compliance assessed.
- Did not run the test suite (CI reports all green).
- Did not verify overlay-mode rendering (only flush-mode logic was reviewed).
- No security, auth, or data-integrity concerns in this change scope.
